nodes/teach_repeat_node.cpp

```diff
@@ -0,0 +1,408 @@
+}
+
+void TeachRepeatNode::playPath(size_t path_index) {
+  const auto& this_path = paths_[path_index];
+
+  auto num_waypoints = this_path.size();
+  auto num_joints = arm_.getGroup()->size();
+  Eigen::MatrixXd wp_matrix(num_joints, num_waypoints);
+  // Set the time based on the recorded times
+  Eigen::VectorXd times(num_waypoints);
+  for (size_t w = 0; w < num_waypoints; ++w) {
+    // Existing joint angles for waypoint
+    auto joints = this_path.waypoint(w);
+    joints = offsetJoints(joints, waypoint_playback_offset_); 
+
+    wp_matrix.col(w) = joints;
+    times(w) = path_dt_ * (w + 1);
+  }
+  
+  // Update stored target position, based on final joint angles.
+  target_xyz_ = arm_.getKinematics().FK(wp_matrix.rightCols<1>());
+  waypoint_playback_offset_ = Eigen::Vector3d::Zero();
+
+  auto t = ::ros::Time::now().toSec();
+  arm_.getTrajectory().replan(t, arm_.getLastFeedback(), wp_matrix, times);
+  ROS_INFO_STREAM("Start path playback through " << num_waypoints << " waypoints");
+}
+
+void TeachRepeatNode::goToWaypoint(const Eigen::VectorXd& joint_angles) {
+  auto t = ::ros::Time::now().toSec();
+
+  auto joints = joint_angles;
+  joints = offsetJoints(joints, waypoint_playback_offset_); 
+
+  target_xyz_ = arm_.getKinematics().FK(joints);
+  arm_.getTrajectory().replan(t, arm_.getLastFeedback(), joints);
+  ROS_INFO_STREAM("Moving to waypoint");
+}
+
+} // namespace ros
+} // namespace hebi
+
+// Shared logic to load a parameter into a variable.  Returns "false" if there is a fatal error
+// (e.g., could not load a required parameter)
+template <typename Param>
+bool loadOptionalParam(ros::NodeHandle& node, Param& param, const Param& default_param, const std::string& param_name) {
+  if (node.hasParam(param_name) && node.getParam(param_name, param)) {
+    ROS_INFO_STREAM("Found and successfully read '" << param_name << "' parameter.");
+    return true;
+  }
+
+  ROS_WARN_STREAM("Could not find/read '" << param_name << "' parameter; setting default value.");
+  param = default_param;
+  return false;
+}
+
+template <typename Param>
+bool loadRequiredParam(ros::NodeHandle& node, Param& param, const std::string& param_name) {
+  if (node.hasParam(param_name) && node.getParam(param_name, param)) {
+    ROS_INFO_STREAM("Found and successfully read '" << param_name << "' parameter.");
+    return true;
+  }
+
+  ROS_ERROR_STREAM("Could not find/read required '" << param_name << "' parameter; aborting!");
+  return false;
+}
+
+int main(int argc, char ** argv) {
+
+  // Initialize ROS node
+  ros::init(argc, argv, "teach_repeat_node");
+  ros::NodeHandle node;
+
+  /////////////////// Load parameters ///////////////////
+
+  // Get parameters for name/family of modules; defaultint to standard values where appropriate
+  std::vector<std::string> families, names;
+  loadOptionalParam(node, families, { "HEBI" }, "families");
+  if (!loadRequiredParam(node, names, "names"))
+    return -1;
```

Comment:
Return codes should always be [positive](https://www.gnu.org/software/libc/manual/html_node/Exit-Status.html)